### PR TITLE
mock-array-big: README.md on Element/constraint.sdc

### PR DESCRIPTION
**Pull request by oharboe**
@maliberty @yupferris @nerriav @tspyrou @maliberty @louiic @torerams @efis1 I've summarized the input I've gotten in the discussions.

Comments?

### CODE REVIEW

**Comment by maliberty** [comment]
It looks fine to me.  @rovinski would you comment as to clarity to someone who wasn't involved in the discussions.

File: flow/designs/asap7/mock-array-big/README.md [hunk 1]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+------------------------------------------------------------------
+
+It isn't possible to articulate a single .sdc file that exactly captures the
+constraints of an Element as each element is used in unique circumstances.
+Instead, a choice has to made and each choice has it's pros and cons.
+
+It is relatively easy to set up a maximum input/output delay for paths
+that start or end in a register, so this case is not discussed here.
+
+It is the combinational(unregistered)
```
Comment (yupferris) [inline flow/designs/asap7/mock-array-big/README.md]:
```suggestion
It is the combinational(unregistered) in -> out
```

File: flow/designs/asap7/mock-array-big/README.md [hunk 2]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+
+In the beginning of a project, during exploration, the .sdc file does not
+need to be particularly detailed. In fact, details can be counterproductive.
+
+Consider minimum input delay, which relates to hold times. If the main concern
+of architectural exploration is to ensure that the design can operate on a
+sufficiently high frequency, then specifying minimum input delays during
+exploration is premature. Minimum input delay is a constraint that belongs at
+the end the development cycle when the design is well established and won't change
+and the concern is to lock down the macro for it's specific operating frequency.
```
Comment (yupferris) [inline flow/designs/asap7/mock-array-big/README.md]:
```suggestion
and the concern is to lock down the macro for its specific operating frequency.
```

File: flow/designs/asap7/mock-array-big/README.md [hunk 3]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+
+Choosing a strategy for Element/constraint.sdc combinational paths
+------------------------------------------------------------------
+
+It isn't possible to articulate a single .sdc file that exactly captures the
+constraints of an Element as each element is used in unique circumstances.
+Instead, a choice has to made and each choice has it's pros and cons.
```
Comment (yupferris) [inline flow/designs/asap7/mock-array-big/README.md]:
```suggestion
Instead, a choice has to made and each choice has its pros and cons.
```

File: flow/designs/asap7/mock-array-big/README.md [hunk 4]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+  However, path segmentation could occur with OpenSTA: https://docs.xilinx.com/r/2020.2-English/ug906-vivado-design-analysis/TIMING-13-Timing-Paths-Ignored-Due-to-Path-Segmentation
+- Overconstrain: set up a maximum input/output path for the element and ignore
+  timing violations at the Element level for combinational paths.
+  The timing violations for the combinational paths are not real violations,
+  because the maximum input and output paths can not occur at the same time
+  in the array. The leftmost Element has the shortest maximum input path and
+  and the rightmost the longestm aximum input path for
+  `assign io_lsbOuts_0 = io_lsbIns_1;`. The OpenROAD tool could put too much
+  emphasis on optimising the combinational path in this case, which could
+  lead to less than optimal results.
```
Comment (yupferris) [inline flow/designs/asap7/mock-array-big/README.md]:
```suggestion
  lead to less than optimal results and/or inflated run times.
```

File: flow/designs/asap7/mock-array-big/README.md [hunk 5]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+being registered.
+
+Clearly, the maximum input delay for `io_lsbIns_1` is
+smaller for the leftmost than for the rightmost Element.
+
+Taking a step back, it is also worth considering what level of detail is
+appropriate for the Element's constraint.sdc file.
+
+In the beginning of a project, during exploration, the .sdc file does not
+need to be particularly detailed. In fact, details can be counterproductive.
```
Comment (yupferris) [inline flow/designs/asap7/mock-array-big/README.md]: Missing a "why" here so I threw this in:

```suggestion
need to be particularly detailed. In fact, details can be counterproductive, as they tend to be inaccurate and cause the tools to spend time solving potential non-issues.
```

**Review by yupferris — APPROVED** [review]
Some nits/suggestions, otherwise lgtm and captures the discussions nicely.

File: flow/designs/asap7/mock-array-big/README.md (hunk 3 above)
Comment (nerriav) [reply flow/designs/asap7/mock-array-big/README.md]: what are the cons against using a unique SDC file for each element ?

File: flow/designs/asap7/mock-array-big/README.md [hunk 6]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+constraints of an Element as each element is used in unique circumstances.
+Instead, a choice has to made and each choice has it's pros and cons.
+
+It is relatively easy to set up a maximum input/output delay for paths
+that start or end in a register, so this case is not discussed here.
+
+It is the combinational(unregistered)
+paths that are tricky to create constraints for.
+
+- Strategy used as of writing: mark non-registered/combinational paths through
```
Comment (nerriav) [inline flow/designs/asap7/mock-array-big/README.md:101]: Timing will be checked but not optimized !!!
'Over- constrained'

File: flow/designs/asap7/mock-array-big/README.md [hunk 7]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+  the Element as false paths. This will not steer the tools to optimize the
+  through element path, such as discussed above and there's no timing information
+  available at the Element level for these combinational paths.
+  However, at the `MockArray` level when the Elements are used, timing will
+  be checked, because the .lib file generated for the Element contains timing
+  information, even for false paths.
+- Use `set_max_delay` for combinational paths. This will make the tool try
+  to optimize the combinational path and there will be timing information available.
+  However, path segmentation could occur with OpenSTA: https://docs.xilinx.com/r/2020.2-English/ug906-vivado-design-analysis/TIMING-13-Timing-Paths-Ignored-Due-to-Path-Segmentation
```
Comment (vijayank88) [inline flow/designs/asap7/mock-array-big/README.md]:
```suggestion
  However, path segmentation could occur with OpenSTA. Refer [here](https://docs.xilinx.com/r/2020.2-English/ug906-vivado-design-analysis/TIMING-13-Timing-Paths-Ignored-Due-to-Path-Segmentation).
```

File: flow/designs/asap7/mock-array-big/README.md [hunk 8]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+- Overconstrain: set up a maximum input/output path for the element and ignore
+  timing violations at the Element level for combinational paths.
+  The timing violations for the combinational paths are not real violations,
+  because the maximum input and output paths can not occur at the same time
+  in the array. The leftmost Element has the shortest maximum input path and
+  and the rightmost the longestm aximum input path for
```
Comment (vijayank88) [inline flow/designs/asap7/mock-array-big/README.md]:
```suggestion
  and the rightmost the longest maximum input path for
```

File: flow/designs/asap7/mock-array-big/README.md (hunk 5 above)
Comment (efis1) [reply flow/designs/asap7/mock-array-big/README.md]: Each step within the implementation flow will be related to different timing "magnitude/concerns", I think we need to segregate (separate by concerns) between different timing concerns within each step of the flow and/vs the tools accuracy.

**Review by vijayank88 — CHANGES_REQUESTED** [review]

File: flow/designs/asap7/mock-array-big/README.md [hunk 9]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+  available at the Element level for these combinational paths.
+  However, at the `MockArray` level when the Elements are used, timing will
+  be checked, because the .lib file generated for the Element contains timing
+  information, even for false paths.
+- Use `set_max_delay` for combinational paths. This will make the tool try
+  to optimize the combinational path and there will be timing information available.
+  However, path segmentation could occur with OpenSTA: https://docs.xilinx.com/r/2020.2-English/ug906-vivado-design-analysis/TIMING-13-Timing-Paths-Ignored-Due-to-Path-Segmentation
+- Overconstrain: set up a maximum input/output path for the element and ignore
+  timing violations at the Element level for combinational paths.
+  The timing violations for the combinational paths are not real violations,
```
Comment (nerriav) [inline flow/designs/asap7/mock-array-big/README.md:113]: It's same incentive of defining the path as 'False path' (there is a strict definition)
Is it a not-real violation ?
what happens when the combinational paths is more complex? (more propagation delay)

File: flow/designs/asap7/mock-array-big/README.md [hunk 10]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+paths that are tricky to create constraints for.
+
+- Strategy used as of writing: mark non-registered/combinational paths through
+  the Element as false paths. This will not steer the tools to optimize the
+  through element path, such as discussed above and there's no timing information
+  available at the Element level for these combinational paths.
+  However, at the `MockArray` level when the Elements are used, timing will
+  be checked, because the .lib file generated for the Element contains timing
+  information, even for false paths.
+- Use `set_max_delay` for combinational paths. This will make the tool try
```
Comment (nerriav) [inline flow/designs/asap7/mock-array-big/README.md:108]: it's something that can be measured by the tool
whether it helps the tool to meet better timing or not

File: flow/designs/asap7/mock-array-big/README.md (hunk 9 above)
Comment (oharboe) [reply flow/designs/asap7/mock-array-big/README.md:113]: I don't have answers to these questions. I'm just summarizing different approaches here. If someone can research and answer these questions, then an update to this README.md is welcome.

File: flow/designs/asap7/mock-array-big/README.md (hunk 2 above)
Comment (efis1) [inline flow/designs/asap7/mock-array-big/README.md]: If we are going to postpone specific timing details per each CE (with regards to its logical structure/content, position within the array and its runtime functionality) and using a generic/symmetric macro per CE, then i suggest to give the tools, at the moment, the freedom for analyzing the comb. path (within the CE) and give the its best results per CE.

File: flow/designs/asap7/mock-array-big/README.md [hunk 11]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+
+In the beginning of a project, during exploration, the .sdc file does not
+need to be particularly detailed. In fact, details can be counterproductive, as they tend to be inaccurate and cause the tools to spend time solving potential non-issues.
+
+Consider minimum input delay, which relates to hold times. If the main concern
+of architectural exploration is to ensure that the design can operate on a
+sufficiently high frequency, then specifying minimum input delays during
+exploration is premature. Minimum input delay is a constraint that belongs at
+the end the development cycle when the design is well established and won't change
+and the concern is to lock down the macro for its specific operating frequency.
```
Comment (efis1) [inline flow/designs/asap7/mock-array-big/README.md:86]: If we are going to postpone specific timing details per each CE (with regards to its logical structure/content, position within the array and its runtime functionality) and using a generic/symmetric macro per CE, then i suggest to give the tools, at the moment, the freedom for analyzing the comb. path (within the CE) and give the its best results per CE.

File: flow/designs/asap7/mock-array-big/README.md [hunk 12]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+Instead, a choice has to made and each choice has its pros and cons.
+
+It is relatively easy to set up a maximum input/output delay for paths
+that start or end in a register, so this case is not discussed here.
+
+It is the combinational(unregistered) in -> out
+paths that are tricky to create constraints for.
+
+- Strategy used as of writing: mark non-registered/combinational paths through
+  the Element as false paths. This will not steer the tools to optimize the
```
Comment (efis1) [inline flow/designs/asap7/mock-array-big/README.md:102]: So are we keeping the false paths?

File: flow/designs/asap7/mock-array-big/README.md [hunk 13]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+  through element path, such as discussed above and there's no timing information
+  available at the Element level for these combinational paths.
+  However, at the `MockArray` level when the Elements are used, timing will
+  be checked, because the .lib file generated for the Element contains timing
+  information, even for false paths.
+- Use `set_max_delay` for combinational paths. This will make the tool try
+  to optimize the combinational path and there will be timing information available.
+  However, path segmentation could occur with OpenSTA. Refer [here](https://docs.xilinx.com/r/2020.2-English/ug906-vivado-design-analysis/TIMING-13-Timing-Paths-Ignored-Due-to-Path-Segmentation).
```
Comment (efis1) [inline flow/designs/asap7/mock-array-big/README.md:110]: FMO we should avoid (as possible) using FPGA references with regards to ASIC issues.

File: flow/designs/asap7/mock-array-big/README.md [hunk 14]
```diff
@@ -33,3 +33,63 @@ Now build the design as usual:
...
+- Overconstrain: set up a maximum input/output path for the element and ignore
+  timing violations at the Element level for combinational paths.
+  The timing violations for the combinational paths are not real violations,
```
Comment (efis1) [inline flow/designs/asap7/mock-array-big/README.md:113]: How does using over constraints initially can help for identify real issues with regards to FE/BE/Tools?

**Comment by oharboe** [comment]
@maliberty I discussed this PR with @efis1, @nerriav and @yupferris offline and we've resolved the issues, so for my part this PR is ready to go.

**Review by vijayank88 — APPROVED** [review]

**Review by maliberty — APPROVED** [review]

**Pull request merged**